dot/network/sync.go

```diff
@@ -18,33 +18,55 @@ package network
 
 import (
 	"context"
-	"errors"
-	"fmt"
-	"reflect"
-	"sort"
-	"sync"
-	"sync/atomic"
 	"time"
 
-	"github.com/ChainSafe/gossamer/dot/types"
-	"github.com/ChainSafe/gossamer/lib/blocktree"
-	"github.com/ChainSafe/gossamer/lib/common"
-	"github.com/ChainSafe/gossamer/lib/common/optional"
-	"github.com/ChainSafe/gossamer/lib/common/variadic"
-
-	"github.com/ChainSafe/chaindb"
 	libp2pnetwork "github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
 )
 
-// SendBlockReqestByHash sends a block request to the network with the given block hash
-func (s *Service) SendBlockReqestByHash(hash common.Hash) {
-	req := createBlockRequestWithHash(hash, blockRequestSize)
-	s.syncQueue.requestDataByHash.Delete(hash)
-	s.syncQueue.trySync(&syncRequest{
-		req: req,
-		to:  "",
-	})
+var maxBlockResponseSize uint64 = 1024 * 1024 * 4 // 4mb
+
+// DoBlockRequest sends a request to the given peer. If a response is received within a certain time period, it is returned, otherwise an error is returned.
+func (s *Service) DoBlockRequest(to peer.ID, req *BlockRequestMessage) (*BlockResponseMessage, error) {
+	fullSyncID := s.host.protocolID + syncID
+
+	s.host.h.ConnManager().Protect(to, "")
+	defer s.host.h.ConnManager().Unprotect(to, "")
+	defer s.host.closeStream(to, fullSyncID)
+
+	// TODO: make this a constant
+	ctx, cancel := context.WithTimeout(s.ctx, time.Second*5)
+	defer cancel()
+
+	stream, err := s.host.h.NewStream(ctx, to, fullSyncID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.host.writeToStream(stream, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.receiveBlockResponse(stream)
+}
+
+func (s *Service) receiveBlockResponse(stream libp2pnetwork.Stream) (*BlockResponseMessage, error) {
+	// TODO: should we create another buffer pool for block response buffers?
+	// for bootstrap this is ok since it's not parallelized, but will need to be updated for tip-mode
+	s.blockResponseBufMu.Lock()
+	defer s.blockResponseBufMu.Unlock()
+
+	buf := s.blockResponseBuf
```

Comment:
allocating a new (large) buffer every time slows down the syncing by a lot, as malloc is one of the most CPU intensive tasks. (this was determined when previously profiling the node). thus we should allocate buffers at startup and re-use them instead of allocating new ones each time. this is why we have the buffer pool, however the buffers there are much smaller than what's needed here, so I just created this fixed buffer for responses for now. we can eventually change it into a buffer pool

Reply:
added a comment to document this!

Reply:
now that makes sense, fine!